core/tee/tee_svc.c

```diff
@@ -597,40 +629,31 @@ static TEE_Result tee_svc_copy_param(struct tee_ta_session *sess,
 		case TEE_PARAM_TYPE_MEMREF_INPUT:
 		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
 		case TEE_PARAM_TYPE_MEMREF_INOUT:
-			if (param->params[n].memref.buffer == NULL) {
-				if (param->params[n].memref.size != 0)
+			va = (void *)param->u[n].mem.offs;
+			s = param->u[n].mem.size;
+			if (!va) {
+				if (!s)
 					return TEE_ERROR_BAD_PARAMETERS;
 				break;
 			}
 			/* uTA cannot expose its private memory */
-			if (tee_mmu_is_vbuf_inside_ta_private(utc,
-				    param->params[n].memref.buffer,
-				    param->params[n].memref.size)) {
+			if (tee_mmu_is_vbuf_inside_ta_private(utc, va, s)) {
 
-				s = ROUNDUP(param->params[n].memref.size,
-						sizeof(uint32_t));
+				s = ROUNDUP(s, sizeof(uint32_t));
 				/* Check overflow */
 				if (req_mem + s < req_mem)
 					return TEE_ERROR_BAD_PARAMETERS;
 				req_mem += s;
 				ta_private_memref[n] = true;
 				break;
 			}
```

Comment:
The old code insured that the targe vbuf was either fully inside TA private or fully outside TA private.

It prevented passing buffers partly inside TA private memory (i.e code/data ends on 4kB boundary and a TA memrefs params starts right at the code/data end boundary).